src/fermilib/utils/_jellium_trotter_error.py

```diff
@@ -0,0 +1,461 @@
+    for operator in all_changes:
+        counts[operator[0]] = counts.get(operator[0], 0) + 2 * operator[1] - 1
+
+    # If the final result creates or destroys the same mode twice.
+    commutes = max(itervalues(counts)) > 1 or min(itervalues(counts)) < -1
+
+    return commutes
+
+
+def jellium_error_operator(terms, indices=None, is_hopping_operator=None,
+                           series_order=2):
+    """Determine the difference between the exact generator of unitary
+    evolution and the approximate generator given by Trotter-Suzuki
+    to the given order.
+
+    Args:
+        terms: a list of FermionOperators in the Hamiltonian
+            to be simulated.
+        indices: a set of indices the terms act on in the same order as terms.
+        is_hopping_operator: a list of whether each term is a hopping operator.
+        series_order: the order at which to compute the BCH expansion.
+            Only the second order formula is currently implemented
+            (corresponding to Equation 9 of the paper).
+
+    Returns:
+        The difference between the true and effective generators of time
+            evolution for a single Trotter step.
+
+    Notes: follows Equation 9 of Poulin et al.'s work in "The Trotter Step
+        Size Required for Accurate Quantum Simulation of Quantum Chemistry".
+    """
+    if series_order != 2:
+        raise NotImplementedError
+    more_info = bool(indices)
+
+    error_operator = FermionOperator.zero()
+    for beta in range(len(terms)):
+        for alpha in range(beta + 1):
+            for alpha_prime in range(beta):
+                # If both are number operators they commute.
+                if more_info and not trivially_double_commutes_more_info(
+                        indices[alpha], indices[beta], indices[alpha_prime],
+                        is_hopping_operator[alpha], is_hopping_operator[beta],
+                        is_hopping_operator[alpha_prime]):
+                    # Determine the result of the double commutator.
+                    double_com = double_commutator(
+                        terms[alpha], terms[beta], terms[alpha_prime],
+                        indices[beta], indices[alpha_prime],
+                        is_hopping_operator[beta],
+                        is_hopping_operator[alpha_prime])
+                    if alpha == beta:
+                        double_com /= 2.0
+
+                    error_operator += double_com
+
+                elif not trivially_double_commutes_jellium(
+                        terms[alpha], terms[beta], terms[alpha_prime]):
+                    double_com = double_commutator(
+                        terms[alpha], terms[beta], terms[alpha_prime])
+
+                    if alpha == beta:
+                        double_com /= 2.0
+
+                    error_operator += double_com
+
+    error_operator /= 12.0
+    return error_operator
+
+
+def jellium_error_bound(terms, indices=None, is_hopping_operator=None,
```

Comment:
I don't like this function, please change. First, it is not specific to jellium. Second, I do not believe we should have a function that takes terms, computers the error operator and then takes the triangle inequality upper bound. That should always be done in two steps by the user. First they compute the error operator with one function, then they call another function that takes the triangle inequality upper bound. You should add a general-purpose function to utils that takes triangle inequality upper-bounds on QubitOperators and FermionOperators. In fact, it should also take InteractionOperators (this is like two more lines).

If you like, you can keep this "loose" bound thing but it is totally general (not even specific to the dual basis) and should be its own function in the general trotter_error module. Furthermore, be careful because I think that might be a bound on the number of trotter steps whereas the "tight" thing is a bound on the error.